### PR TITLE
Predict unit tests

### PR DESCRIPTION
**Pull request by lilly-muyama**
# Unit tests for the Predict Microservice

**_WHAT DOES THIS PR DO?_**
This PR creates unit tests for the predict microservice

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
https://airqoteam.atlassian.net/browse/PLAT-432

**_WHAT IS THE LINK TO THE PR BRANCH_**

**_HOW DO I TEST OUT THIS PR?_**
1. Git clone this repo
2. `cd src/predict` 
3.  `env\scripts\activate` for Windows and `source env/bin/activate` for Linux
4. `pip install -r requirements.txt`
5. To run all test files: run `pytest`
6. To run a specific file only, run `pytest file_name.py`
7. To run a specific test function, run `pytest file_name.py::function_name`

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 
N/A

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**

**_ARE THERE ANY RELATED PRs?_**

### CODE REVIEW

File: src/predict/tests/test_datamanagement.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
+import pytest
+import sys
+sys.path.append('./')
+from models.datamanagement import get_channel_data_raw, save_predictions, get_channel_best_configurations
+from models.datamanagement import calculate_hourly_averages, get_channel_id, get_all_coordinates
```
Comment (Baalmart) [inline src/predict/tests/test_datamanagement.py:5]: For unit tests, could we mock the data and not make a direct call to the actual database? More notes here:[ 1](https://stackoverflow.com/questions/51988331/using-a-fake-mongodb-for-pytest-testing) and [2](https://pypi.org/project/pytest-mongodb/)

**Review by Baalmart — COMMENTED** [review]
Thanks @lilly-muyama , looking good. :) 
A few thoughts/questions in addition to another specific review comment in the PR follow:

1. Could we have the test files alongside the files they are testing? WHY? For easy navigation during refactoring works in the near future.
2. Is there a possibility of categorising/grouping/tagging our tests so that a person can specify which ones to run?

**Comment by bobturneruk** [comment]
Looks like there's a lot of excellent work here. I'm trying to run on my machine and hit this error on pip install:

```
ERROR: Cannot install -r requirements.txt (line 28), -r requirements.txt (line 31) and google-resumable-media==0.5.1 because these package versions have conflicting dependencies.

The conflict is caused by:
    The user requested google-resumable-media==0.5.1
    google-cloud-bigquery 1.25.0 depends on google-resumable-media<0.6dev and >=0.5.0
    google-cloud-storage 1.30.0 depends on google-resumable-media<2.0dev and >=0.6.0

To fix this you could try to:
1. loosen the range of package versions you've specified
2. remove package versions to allow pip attempt to solve the dependency conflict

ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
```

I'm using `conda` for virtual environments, but I think this is likely to be a conflict that affects all virtual environments. Could the version requirements be relaxed @lilly-muyama?

**Comment by lilly-muyama** [comment]
> 
> 
> Looks like there's a lot of excellent work here. I'm trying to run on my machine and hit this error on pip install:
> 
> ```
> ERROR: Cannot install -r requirements.txt (line 28), -r requirements.txt (line 31) and google-resumable-media==0.5.1 because these package versions have conflicting dependencies.
> 
> The conflict is caused by:
>     The user requested google-resumable-media==0.5.1
>     google-cloud-bigquery 1.25.0 depends on google-resumable-media<0.6dev and >=0.5.0
>     google-cloud-storage 1.30.0 depends on google-resumable-media<2.0dev and >=0.6.0
> 
> To fix this you could try to:
> 1. loosen the range of package versions you've specified
> 2. remove package versions to allow pip attempt to solve the dependency conflict
> 
> ERROR: ResolutionImpossible: for help visit https://pip.pypa.io/en/latest/user_guide/#fixing-conflicting-dependencies
> ```
> 
> I'm using `conda` for virtual environments, but I think this is likely to be a conflict that affects all virtual environments. Could the version requirements be relaxed @lilly-muyama?

@bobturneruk I am currently making many updates to this PR and I will look into this too. Thanks.

**Comment by bobturneruk** [comment]
Great. Thanks. I'm keen to help. Another suggestion is to implement some CI as you're going along (based on this https://github.com/airqo-platform/AirQo-modules/blob/staging/.github/workflows/python-app.yml but without `tox`). This will help ensure that the tests pass on a clean install without user intervention. See also https://github.com/pykale/pykale/blob/master/.github/workflows/test.yml

**Review by ja-odur — CHANGES_REQUESTED** [review]
The failing tests are mainly due to unhandled exceptions in the codebase, It will be better to actually test for these specific exceptions (ensuring the tests pass). In the future when these exceptions are handled, the tests will fail ( and then they can be fixed by the "implementer")

**Comment by sserurich** [comment]
> The failing tests are mainly due to unhandled exceptions in the codebase, It will be better to actually test for these specific exceptions (ensuring the tests pass). In the future when these exceptions are handled, the tests will fail ( and then they can be fixed by the "implementer")

image
This is comment applies to what I get.

**Pull request closed without merging**